### PR TITLE
*: compact with physical in client side

### PR DESCRIPTION
**Pull request by gyuho**
Related https://github.com/coreos/etcd/pull/5700.

We need to freeze v3 today in a separate branch. Since it's variadic, it won't really break anything, but it's still an interface change, so we should put this in v3 if we are going to change.

### CODE REVIEW

File: clientv3/compact_op.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+// CompactOp represents a compact Operation.
+type CompactOp struct {
+	revision int64
+	physical bool
+}
+
+// CompactOpOption configures compact operation.
+type CompactOpOption func(*CompactOp)
```
Comment (heyitsanthony) [inline clientv3/compact_op.go]: just `CompactOption`?

File: etcdctl/ctlv3/command/compaction_command.go [hunk 2]
```diff
@@ -18,16 +18,21 @@ import (
...
+
 // NewCompactionCommand returns the cobra command for "compaction".
 func NewCompactionCommand() *cobra.Command {
-	return &cobra.Command{
+	cmd := &cobra.Command{
 		Use:   "compaction <revision>",
 		Short: "Compaction compacts the event history in etcd.",
 		Run:   compactionCommandFunc,
 	}
+	cmd.Flags().BoolVar(&compactPhysical, "physical", false, "'true' to wait until compaction is physically applied to the local database.")
```
Comment (heyitsanthony) [inline etcdctl/ctlv3/command/compaction_command.go]: wait for compaction to physically remove all old revisions

File: etcdctl/ctlv3/command/compaction_command.go [hunk 3]
```diff
@@ -41,6 +46,11 @@ func compactionCommandFunc(cmd *cobra.Command, args []string) {
 		ExitWithError(ExitError, err)
 	}
 
+	var opts []clientv3.CompactOpOption
+	if compactPhysical {
+		opts = append(opts, clientv3.WithCompactPhysical())
```
Comment (heyitsanthony) [inline etcdctl/ctlv3/command/compaction_command.go]: this is never used?

File: clientv3/compact_op.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clientv3
+
+import (
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+)
+
+// CompactOp represents a compact Operation.
```
Comment (heyitsanthony) [inline clientv3/compact_op.go]: s/Op/op

**Comment by gyuho** [comment]
@heyitsanthony All addressed and tests passed. PTAL.

Thanks!

**Comment by heyitsanthony** [comment]
lgtm, defer to @xiang90

**Comment by xiang90** [comment]
@gyuho ci failed.

**Comment by gyuho** [comment]
@xiang90 Just created a separate issue here https://github.com/coreos/etcd/issues/5783.

**Comment by xiang90** [comment]
can you re-trigger the test?

**Comment by gyuho** [comment]
@xiang90 PTAL. Tests just passed. Thanks.

**Comment by xiang90** [comment]
lgtm

**Comment by mqliang** [comment]
@gyuho @xiang90 

Make the compaction interface as `Compact(context.Context, *CompactionRequest) (*CompactionResponse, error)` instead? Since the proxy must implement the [KVServer](https://github.com/coreos/etcd/blob/master/etcdserver/etcdserverpb/rpc.pb.go#L2137-L2157) interface, we need return `*CompactionResponse`.

**Pull request merged**